### ISSUE TITLE
Converting `TypedDict` to another `TypedDict`, Pyright doesn't handle it correctly.

### ISSUE THREAD

**Issue by NiKuma0** [microsoft/pyright#8646]
**Describe the bug**
When converting `TypedDict` to another `TypedDict`, Pyright doesn't handle it correctly:

If one `TypedDict` is similar to another but has an extra key, you **can** convert one to the other without a Pyright error. 
If one `TypedDict` has no `id` and another has it but is marked as `NotRequired`, you **can't** convert one to the other without a Pyright error (`id` is missing).

**Code or Screenshots**

```python
import typing as t


class Entity(t.TypedDict):
    id: t.NotRequired[int]
    username: str

class EntityToCreate(t.TypedDict):
    username: str

class EntityToService(t.TypedDict):
    username: str
    extra_data: str

def service(**kwargs: t.Unpack[EntityToService]):
    repository(**EntityToCreate(kwargs))  # Have no errors, even that `EntityToService` has `extra_data`.

def repository(**kwargs: t.Unpack[EntityToCreate]):
    create(**Entity(kwargs))  # Have error (`id` is missing), even though `id` is `NotRequired` - error doesn't seem right.

def create(**kwargs: t.Unpack[Entity]): ...
```
Pyright logs:
```
../test.py
  ../test.py:19:21 - error: Argument of type "*EntityToCreate" cannot be assigned to parameter "__map" of type "Entity" in function "__init__"
    "id" is missing from "*EntityToCreate" (reportArgumentType)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
```
pyright --version
pyright 1.1.374
```
```
code --version
1.92.0
b1c0a14de1414fcdaa400695b4db1c0799bc3124
x64
```

**Comment by erictraut** [comment]
Pyright's behavior is correct here, so I don't consider this a bug. Subtype relationships of `TypedDicts` are well defined [in the typing spec](https://typing.readthedocs.io/en/latest/spec/typeddict.html) and the corresponding [conformance tests](https://github.com/python/typing/blob/main/conformance/tests/typeddicts_inheritance.py).

You cannot assign a value of type `EntityToCreate` to `Entity` even though `id` is marked `NotRequired`. The reason is that a value of type `EntityToCreate` can contain additional keys that are not specified. This could include an `id` key whose value type would violate the definition of `EntityToCreate`.

Not surprising, mypy and other conformant type checkers also generate a type violation error if you attempt to assign an `EntityToCreate` to a variable or parameter that requires an `Entity`.

**Comment by NiKuma0** [comment]
@erictraut but still, why I can convert `EntityToService` to `EntityToCreate` without errors?

> You cannot assign a value of type `EntityToCreate` to `Entity` even though `id` is marked `NotRequired`. The reason is that a value of type `EntityToCreate` can contain additional keys that are not specified. This could include an `id` key whose value type would violate the definition of `EntityToCreate`.

Any `TypedDict` cannot contain additional keys (without error), so I don't really understand this point.
```python
repository(username='username', notspec='notspec')  # No parameter named "notspec" (reportCallIssue)

entity: EntityToCreate = {'username': 'username', 'addspec': 'addspec'}
# Expression of type "dict[str, str]" is incompatible with declared type "EntityToCreate"
    # "addspec" is an undefined item in type "EntityToCreate" (reportAssignmentType)
```

`TypedDict` is a structural type, so it should assignable to another if the only difference is that one contained `id: NotRequired[int]` and another does not. There is no inconsistencies with [PEP8](https://typing.readthedocs.io/en/latest/spec/typeddict.html#assignability).

**Comment by erictraut** [comment]
You can convert `EntityToService` to `EntityToCreate` without errors because these are structural types, and any object that meets the structural requirements of `EntityToService` will also meet the structural requirements of `EntityToCreate`.

That cannot be said when converting `EntityToCreate` to `Entity`.

> Any TypedDict cannot contain additional keys (without error), so I don't really understand this point.

I don't understand this statement. `TypedDict` is a structural type definition; a type that matches the structural definition can have additional keys beyond those that are specified in the `TypedDict`.

Consider if I add the following code to the bottom of your example:

```python
class EntityWithStrId(t.TypedDict):
    id: str
    username: str

a = EntityWithStrId(id="x", username="john")
b: EntityToCreate = a  # No type error
c: Entity = b  # Type error
```
